src/compiler/checker.ts

```diff
@@ -38974,15 +38978,34 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
                 else if (contextualSignature && !node.typeParameters && contextualSignature.parameters.length > node.parameters.length) {
                     const inferenceContext = getInferenceContext(node);
-                    if (checkMode && checkMode & CheckMode.Inferential) {
+                    if (checkMode & CheckMode.Inferential) {
                         inferFromAnnotatedParameters(signature, contextualSignature, inferenceContext!);
                     }
                 }
                 if (contextualSignature && !getReturnTypeFromAnnotation(node) && !signature.resolvedReturnType) {
-                    const returnType = getReturnTypeFromBody(node, checkMode);
-                    if (!signature.resolvedReturnType) {
-                        signature.resolvedReturnType = returnType;
+                    let contextualReturnType: Type;
+                    let returnType: Type;
+
+                    if (node.typeParameters) {
+                        checkMode |= CheckMode.SkipReturnTypeFromBodyInference;
+                    }
+
+                    if (
+                        isNodeContextSensitive && ((checkMode & (CheckMode.Inferential | CheckMode.SkipReturnTypeFromBodyInference)) === CheckMode.Inferential) &&
+                        couldContainTypeVariables(contextualReturnType = getReturnTypeOfSignature(contextualSignature))
+                    ) {
+                        const inferenceContext = getInferenceContext(node);
+                        const isReturnContextSensitive = !!node.body && (node.body.kind === SyntaxKind.Block ? forEachReturnStatement(node.body as Block, statement => !!statement.expression && isContextSensitive(statement.expression)) : isContextSensitive(node.body));
+                        returnType = getReturnTypeFromBody(node, checkMode | (isReturnContextSensitive ? CheckMode.SkipContextSensitive : 0));
+                        inferTypes(inferenceContext!.inferences, returnType, contextualReturnType);

```

Comment:
alternatively, this could be handled by intra expression inference **but** that is order-sensitive and traditionally simple cases like this one aren't:
```ts
declare function test<T>(_: {
  stuff: T,
  consume: (arg: T) => void
}): void

test({
  consume: (arg) => {},
  stuff: 'foo' // this can come after `consume`
})
```

So the reason I put this logic *here* is that it allows for the same order-independence